flow-data/src/main/java/com/vaadin/flow/data/binder/DefaultBindingExceptionHandler.java

```
@@ -0,0 +1,108 @@
+            Exception exception) {
+        if (!(field instanceof HasElement)) {
+            return Optional.empty();
+        }
+        Element element = ((HasElement) field).getElement();
+        StringBuilder majorProperties = new StringBuilder();
+        String label = element.getProperty(LABEL);
+        if (label != null) {
+            appendProperty(majorProperties, LABEL, label);
+        }
+        String id = element.getProperty(ID);
+        if (id != null) {
+            appendProperty(majorProperties, ID, id);
+        }
+
+        UI ui = getUI(element);
+        if (majorProperties.length() == 0 && ui != null
+                && ui.getSession() != null
+                && !ApplicationConfiguration
+                        .get(ui.getSession().getService().getContext())
+                        .isProductionMode()) {
+            element.getAttributeNames()
+                    .forEach(attribute -> appendProperty(majorProperties,
+                            attribute, element.getAttribute(attribute)));
+            element.getPropertyNames()
+                    .forEach(property -> appendProperty(majorProperties,
+                            property, element.getProperty(property)));
+        }
+
+        if (majorProperties.length() > 0) {
+            majorProperties.insert(0,
+                    "An exception has been thrown inside binding logic for the field element [")
+                    .append("]");
+        }
+
+        if (majorProperties.length() > 0) {
+            return Optional.of(new BindingException(majorProperties.toString(),
+                    exception));
+        }
+        return Optional.empty();
+    }
+
+    private UI getUI(Element element) {
+        NodeOwner owner = element.getNode().getOwner();
+        UI ui = null;
+        if (element.getNode().isAttached()) {
+            ui = ((StateTree) owner).getUI();
+        }
+        if (ui == null) {
+            ui = UI.getCurrent();
```

I don't get the meaning of traversing elements tree, if it finally falls into `ui = UI.getCurrent();` anyway.
Could you please explain why you don't just use `UI.getCurrent();` instead of `getUI`?

---

Thanks for explanations.
Due to the method names are the same - `getUI`, I erroneously concluded that the method calls itself recursively but in fact it calls `getUI` of the owner.
 
> I would recommend to avoid questions unrelated to the review in the review. I see too many of them recently and they are blocking the PR unnecessary.

For me it's a new code and I don't always know the patterns used in the framework.
I believe it's better to ask rather that silently skip it, not having the idea how it works.

---

I don't agree with this since the purpose of the review is not to learn the code by the reviewer.
It's just a wrong usage of the tool.
It's OK when the code is complicated and when it happens rarely.
But it happens regularly. And it's not that complicated since it's possible to spend some time and find out the answer by yourself via reading and checking the code: it's available.